File: src/main/java/bowling/Game.java
package bowling;

public class Game {
    private int currentRoll;
    private int[] rolls = new int[21];

    public void roll(int pins) {
        rolls[currentRoll++] = pins;
    }

    public int score() {
        int score = 0;
        int frameIndex = 0;
        for (int frame = 0; frame < 10; frame++) {
            if (isStrike(frameIndex)) {
                score += 10 + strikeBonus(frameIndex);
                frameIndex += 1;
            } else if (isSpare(frameIndex)) {
                score += 10 + spareBonus(rolls[frameIndex + 2]);
                frameIndex += 2;
            } else {
                score += pinsThrownInStrike(frameIndex);
                frameIndex += 2;
            }
        }
        return score;
    }

    private boolean isStrike(int frameIndex) {
        return rolls[frameIndex] == 10;
    }

    private int spareBonus(int roll) {
        return roll;
    }

    private int strikeBonus(int frameIndex) {
        return rolls[frameIndex + 1] + rolls[frameIndex + 2];
    }

    private int pinsThrownInStrike(int frameIndex) {
        return rolls[frameIndex] + rolls[frameIndex + 1];
    }

    private boolean isSpare(int frameIndex) {
        return rolls[frameIndex] + rolls[frameIndex + 1] == 10;
    }
}

File: src/main/ruby/bowling/game.rb
class Game
  def initialize
    @current_roll = 0
    @rolls = []
  end

  def roll(pins)
    @rolls[@current_roll] = pins
    @current_roll += 1
  end

  def score
    score = frame_index = 0
    10.times do
      if strike?(frame_index)
        score += 10 + strike_bonus(frame_index)
        frame_index += 1
      elsif spare?(frame_index)
        score += 10 + spare_bonus(@rolls[frame_index + 2])
        frame_index += 2
      else
        score += 10 + pins_thrown_in_strike(frame_index)
        frame_index += 2
      end
    end
    score
  end

  def strike?(frame_index)
    @rolls[frame_index] == 10
  end

  def spare?(frame_index)
    @rolls[frame_index].to_i + @rolls[frame_index + 1].to_i == 10
  end

  def spare_bonus(roll)
    roll
  end

  def strike_bonus(frame_index)
    @rolls[frame_index + 1].to_i + @rolls[frame_index + 2].to_i
  end

  def pins_thrown_in_strike(frame_index)
    @rolls[frame_index].to_i + @rolls[frame_index + 1].to_i
  end
end
